tests/CredentialSource/ExecutableSourceTest.php

```diff
@@ -0,0 +1,328 @@
+     */
+    public function testCachedTokenFile()
+    {
+        putenv('GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES=1');
+
+        $outputFile = tempnam(sys_get_temp_dir(), 'token');
+        file_put_contents($outputFile, json_encode([
+            'version' => 1,
+            'success' => true,
+            'token_type' => 'urn:ietf:params:oauth:token-type:id_token',
+            'id_token' => 'abc',
+            'expiration_time' => time() + 100,
+        ]));
+
+        $source = new ExecutableSource('fake-command', $outputFile);
+        $subjectToken = $source->fetchSubjectToken();
+        $this->assertEquals('abc', $subjectToken);
+    }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testCachedTokenFileExpiredCallsExecutable()
+    {
+        putenv('GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES=1');
+
+        $cachedToken = [
+            'version' => 1,
+            'success' => true,
+            'token_type' => 'urn:ietf:params:oauth:token-type:id_token',
+            'id_token' => 'abc',
+            // token is expired
+            'expiration_time' => time() - 100,
+        ];
+        $successToken = ['expiration_time' => time() + 100] + $cachedToken;
+        $outputFile = tempnam(sys_get_temp_dir(), 'token');
+        file_put_contents($outputFile, json_encode($cachedToken));
+
+        $executableHandler = $this->prophesize(ExecutableHandler::class);
+        $executableHandler->__invoke('fake-command')
+            ->shouldBeCalledOnce()
+            ->willReturn(0);
+        $executableHandler->getOutput()
+            ->shouldBeCalledOnce()
+            ->willReturn(json_encode($successToken));
+
+        $source = new ExecutableSource('fake-command', $outputFile, $executableHandler->reveal());
+        $subjectToken = $source->fetchSubjectToken();
+        $this->assertEquals('abc', $subjectToken);
+    }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testCachedTokenFileWithSuccessFalseCallsExecutable()
+    {
+        putenv('GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES=1');
+
+        $cachedToken = [
+            'version' => 1,
+            // token has success=false
+            'success' => false,
+            'code' => 0,
+            'message' => 'error!'
+        ];
+        $successToken = [
+            'version' => 1,
+            'success' => true,
+            'token_type' => 'urn:ietf:params:oauth:token-type:id_token',
+            'id_token' => 'abc',
+            'expiration_time' => time() + 100,
+        ];
+        $outputFile = tempnam(sys_get_temp_dir(), 'token');
+        file_put_contents($outputFile, json_encode($cachedToken));
+
+        $executableHandler = $this->prophesize(ExecutableHandler::class);
+        $executableHandler->__invoke('fake-command')
+            ->shouldBeCalledOnce()
+            ->willReturn(0);
+        $executableHandler->getOutput()
+            ->shouldBeCalledOnce()
+            ->willReturn(json_encode($successToken));
```

Comment:
This code is being used thrice in this file, we can just export them to a private method the reuse this.

Reply:
This is definitely a matter of preference, but I'd prefer to keep each test defining it's own mocks, rather than abstract parts out to custom functions. The way it is now is easier to follow IMHO